release/kdm/rke2.go

```diff
@@ -0,0 +1,467 @@
+	// to get the latest previous minor.
+	if patch == 0 {
+		prevVersion, err := u.getRKE2LatestMinor(major, minor)
+		if err != nil {
+			return "", err
+		}
+		return prevVersion, nil
+	}
+
+	return fmt.Sprintf(rke2VersionTemplate, major, minor, patch-1, 1), nil
+}
+
+func (u *RKE2ChannelsUpdater) getRKE2LatestMinor(major, minor int) (string, error) {
+	baseVersion := fmt.Sprintf("v%d.%d", major, minor)
+
+	for i := len(u.currentVersions) - 1; i >= 0; i-- {
+		if strings.Contains(u.currentVersions[i], baseVersion) {
+			return u.currentVersions[i], nil
+		}
+	}
+
+	return "", errors.New("not found latest patch for " + baseVersion)
+}
+
+// parseRKE2Version receives a version in this format: vX.Y.Z+rke2rN
+// and returns the major, minor, patch, and release numbers as integers.
+func parseRKE2Version(version string) (int, int, int, int, error) {
+	formatErr := fmt.Errorf("version %q is not in the expected format vX.Y.Z+rke2rN", version)
+
+	if !strings.HasPrefix(version, "v") {
+		return 0, 0, 0, 0, formatErr
+	}
+
+	version = strings.TrimPrefix(version, "v")
+
+	// "1.2.3+rke2r4" -> ["1.2.3", "4"]
+	parts := strings.Split(version, "+rke2r")
+	if len(parts) != 2 {
+		return 0, 0, 0, 0, formatErr
+	}
+	versionStr := parts[0]
+	releaseStr := parts[1]
+
+	// "1.2.3" -> ["1", "2", "3"]
+	versionParts := strings.Split(versionStr, ".")
+	if len(versionParts) != 3 {
+		return 0, 0, 0, 0, formatErr
+	}
+
+	majorStr := versionParts[0]
+	major, err := strconv.Atoi(majorStr)
+	if err != nil {
+		return 0, 0, 0, 0, fmt.Errorf("invalid major version part %q: %w", majorStr, err)
+	}
+
+	minorStr := versionParts[1]
+	minor, err := strconv.Atoi(minorStr)
+	if err != nil {
+		return 0, 0, 0, 0, fmt.Errorf("invalid minor version part %q: %w", minorStr, err)
+	}
+
+	patchStr := versionParts[2]
+	patch, err := strconv.Atoi(patchStr)
+	if err != nil {
+		return 0, 0, 0, 0, fmt.Errorf("invalid patch version part %q: %w", patchStr, err)
+	}
+
+	release, err := strconv.Atoi(releaseStr)
+	if err != nil {
+		return 0, 0, 0, 0, fmt.Errorf("invalid release version part %q: %w", releaseStr, err)
+	}
+
+	return major, minor, patch, release, nil
+}
+
+func (u *RKE2ChannelsUpdater) addRelease(release Release) error {
+	newReleaseContent := make([]*yaml.Node, 0)
+	newReleaseContent = append(newReleaseContent, createScalarNode("version"), createScalarNode(release.Version))
+
+	prevReleasePos, prevRelease, err := u.getPreviousRelease(release.Version)
+	if err != nil {
+		return err
+	}
+
+	newReleaseContent = append(newReleaseContent, createScalarNode("minChannelServerVersion"), createScalarNode(prevRelease.MinChannelServerVersion))
+	newReleaseContent = append(newReleaseContent, createScalarNode("maxChannelServerVersion"), createScalarNode(prevRelease.MaxChannelServerVersion))
+
+	sanitizedVersionForAnchor := strictlyAlphanumeric(release.Version) // e.g., "v1216rke2r1"
+	versionForAnchor := getAnchorName(release.Version)
+
+	// defining charts
```

Comment:
Why are we creating inner scope here and below?

Reply:
I added these scopes just for readability.